File: training.py
import sys
import os
from pyspark.sql import SparkSession
from pyspark.ml import Pipeline, PipelineModel
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.feature import VectorAssembler, StringIndexer
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
from pyspark.sql.functions import col

def clean_data(data_frame):
    """Cleans data by casting columns to double and stripping extra quotes."""
    return data_frame.select(*(col(column).cast("double").alias(column.strip('\"')) for column in data_frame.columns))

if __name__ == "__main__":
    print("Starting Spark Application")
    spark = SparkSession.builder.appName("WineQualityPrediction").getOrCreate()
    spark.sparkContext.setLogLevel('ERROR')

    # Configuration for using S3
    spark._jsc.hadoopConfiguration().set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")

    training_data_path = "s3://path/to/TrainingDataset.csv"
    model_output_path = "s3://path/to/spark-model"

    try:
        print(f"Loading training data from {training_data_path}")
        raw_data_frame = spark.read.format("csv").option('header', 'true').option("sep", ";").option("inferschema", 'true').load(training_data_path)
        training_data_frame = clean_data(raw_data_frame)

        feature_columns = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
                           'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
                           'pH', 'sulphates', 'alcohol']
        label_column = 'quality'

        assembler = VectorAssembler(inputCols=feature_columns, outputCol='features')
        indexer = StringIndexer(inputCol=label_column, outputCol="label")

        rf_classifier = RandomForestClassifier(labelCol='label', featuresCol='features', numTrees=100, maxDepth=10, seed=42)
        
        pipeline = Pipeline(stages=[assembler, indexer, rf_classifier])

        # Set up CrossValidator
        param_grid = ParamGridBuilder() \
            .addGrid(rf_classifier.maxDepth, [5, 10, 15]) \
            .addGrid(rf_classifier.numTrees, [50, 100, 150]) \
            .build()

        evaluator = MulticlassClassificationEvaluator(labelCol='label', predictionCol='prediction', metricName='accuracy')

        cv = CrossValidator(estimator=pipeline, estimatorParamMaps=param_grid, evaluator=evaluator, numFolds=5)
        cv_model = cv.fit(training_data_frame)

        # Select the best model
        best_model = cv_model.bestModel
        print(f"Best model parameters: {best_model.stages[-1].extractParamMap()}")

        # Save the best model
        print(f"Saving the best model to {model_output_path}")
        best_model.write().overwrite().save(model_output_path)

    except Exception as e:
        print(f"An error occurred: {e}")

    spark.stop()

File: training.java
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.ml.Pipeline;
import org.apache.spark.ml.PipelineModel;
import org.apache.spark.ml.classification.RandomForestClassifier;
import org.apache.spark.ml.feature.VectorAssembler;
import org.apache.spark.ml.feature.StringIndexer;
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator;
import org.apache.spark.ml.tuning.CrossValidator;
import org.apache.spark.ml.tuning.ParamGridBuilder;
import org.apache.spark.mllib.evaluation.MulticlassMetrics;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

public class WineQualityPrediction {
    public static void main(String[] args) {
        System.out.println("Starting Spark Application");

        SparkSession spark = SparkSession.builder()
                .appName("WineQualityPrediction")
                .config("spark.some.config.option", "config-value")
                .getOrCreate();

        spark.sparkContext().setLogLevel("ERROR");
        String trainingDataPath = "TrainingDataset.csv";
        String modelOutputPath = "spark-model";

        try {
            System.out.println("Loading training data from " + trainingDataPath);
            Dataset<Row> rawDataFrame = spark.read()
                    .format("csv")
                    .option("header", "true")
                    .option("sep", ";")
                    .option("inferschema", "true")
                    .load(trainingDataPath);

            Dataset<Row> trainingDataFrame = cleanData(rawDataFrame);
            String[] featureColumns = new String[]{"fixed acidity", "volatile acidity", "citric acid", "residual sugar", "chlorides", "free sulfur dioxide", "total sulfur dioxide", "density", "pH", "sulphates", "alcohol"};
            String labelColumn = "quality";
            VectorAssembler assembler = new VectorAssembler()
                    .setInputCols(featureColumns)
                    .setOutputCol("features");
            StringIndexer indexer = new StringIndexer()
                    .setInputCol(labelColumn)
                    .setOutputCol("label");
            RandomForestClassifier rfClassifier = new RandomForestClassifier()
                    .setLabelCol("label")
                    .setFeaturesCol("features")
                    .setNumTrees(100)
                    .setMaxDepth(10)
                    .setSeed(42);

            Pipeline pipeline = new Pipeline()
                    .setStages(new org.apache.spark.ml.PipelineStage[]{assembler, indexer, rfClassifier});
            ParamGridBuilder paramGridBuilder = new ParamGridBuilder()
                    .addGrid(rfClassifier.maxDepth(), new int[]{5, 10, 15})
                    .addGrid(rfClassifier.numTrees(), new int[]{50, 100, 150});
            org.apache.spark.ml.tuning.ParamMap[] paramGrid = paramGridBuilder.build();

            MulticlassClassificationEvaluator evaluator = new MulticlassClassificationEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("accuracy");

            CrossValidator cv = new CrossValidator()
                    .setEstimator(pipeline)
                    .setEstimatorParamMaps(paramGrid)
                    .setEvaluator(evaluator)
                    .setNumFolds(5);

            PipelineModel cvModel = cv.fit(trainingDataFrame);
            PipelineModel bestModel = (PipelineModel) cvModel.bestModel();
            System.out.println("Best model parameters: " + bestModel.stages()[bestModel.stages().length - 1].extractParamMap());
            System.out.println("Saving the best model to " + modelOutputPath);
            bestModel.write().overwrite().save(modelOutputPath);

        } catch (Exception e) {
            System.out.println("An error occurred: " + e.getMessage());
        }

        System.out.println("Exiting Spark Application");
        spark.stop();
    }

    private static Dataset<Row> cleanData(Dataset<Row> dataFrame) {
        for (String column : dataFrame.columns()) {
            dataFrame = dataFrame.withColumn(column.replace("\"", ""), dataFrame.col(column).cast("double"));
        }
        return dataFrame;
    }
}

